sdk/test/trace/tracer_test.cc

```diff
@@ -223,18 +226,105 @@ TEST(Tracer, GetSampler)
+TEST(Tracer, ComprehensiveTest)
 {
-  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received(
+  // Create tracers with AlwaysOn, AlwaysOff, Parent-or-else, and Probability samplers.
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_on(
       new std::vector<std::unique_ptr<SpanData>>);
-  auto tracer = initTracer(spans_received);
-
-  auto span = tracer->StartSpan("span 1");
-
-  span->SetAttribute("abc", 3.1);
-
-  span->End();
-  ASSERT_EQ(1, spans_received->size());
-  auto &span_data = spans_received->at(0);
-  ASSERT_EQ(3.1, nostd::get<double>(span_data->GetAttributes().at("abc")));
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_off(
+      new std::vector<std::unique_ptr<SpanData>>);
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_parent_on(
+      new std::vector<std::unique_ptr<SpanData>>);
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_probability_on(
+      new std::vector<std::unique_ptr<SpanData>>);
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_probability_off(
+      new std::vector<std::unique_ptr<SpanData>>);
+  auto tracer_on  = initTracer(spans_received_on);
+  auto tracer_off = initTracer(spans_received_off, std::make_shared<AlwaysOffSampler>());
+  // Current ShouldSample always pass an empty ParentContext,
+  // so this sampler will work as an AlwaysOnSampler.
+  auto tracer_parent_on =
+      initTracer(spans_received_parent_on,
+                 std::make_shared<ParentOrElseSampler>(std::make_shared<AlwaysOnSampler>()));
+  // Current ShouldSample uses an invalid TraceId.
+  // This sampler always return RECORD_AND_SAMPLE at probability 0.01
+  auto tracer_probability_on =
+      initTracer(spans_received_probability_on, std::make_shared<ProbabilitySampler>(0.01));
+  // This sampler always return NOT_SAMPLE at probability 0
+  auto tracer_probability_off =
+      initTracer(spans_received_probability_off, std::make_shared<ProbabilitySampler>(0));
+
+  // Testing AlwaysOn sampler.
+  // Create two spans for each tracer. Check the exported result.
+  auto span_on_1 = tracer_on->StartSpan("span with AlwaysOn sampler 1");
+  auto span_on_2 = tracer_on->StartSpan("span with AlwaysOn sampler 2");
+
+  span_on_1->SetAttribute("attr1", 3.1);
+
+  span_on_2->End();
+  span_on_1->End();
+  ASSERT_EQ(2, spans_received_on->size());
+  auto &span_data_on_1 = spans_received_on->at(1);  // span 2 ends first.
+  auto &span_data_on_2 = spans_received_on->at(0);
+  ASSERT_EQ(3.1, nostd::get<double>(span_data_on_1->GetAttributes().at("attr1")));
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_1->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_1->GetDuration());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_2->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_2->GetDuration());
```

Comment:
This code seems to be repeated for every Tracer. Can you pull this out into a helper function?

Reply:
Replaced these check to simply checking the span name.